Add to naginator reason a clickable link to original failed build

litmanovich opened the pull request:
Make the number of build which caused retry a clickable link

KostyaSha commented:
Will it work with enabled jenkins security for html escape markup?

ikedam commented:
* The text will be html-escaped by Jenkins when displayed.
    * https://github.com/jenkinsci/jenkins/blob/jenkins-1.642/core/src/main/resources/hudson/model/Cause/description.jelly
    * Would you attatch a screenshot to see how this change works?
* The link should be disabled when the source build is deleted.
    * like `UpstreamCause`: https://github.com/jenkinsci/jenkins/blob/jenkins-1.642/core/src/main/resources/hudson/model/Cause/UpstreamCause/description.jelly

litmanovich commented:
The second commit addresses the case of source build deleted.

Screenshot of naginator cause with link: 
![naginatorcauselink](https://cloud.githubusercontent.com/assets/1868117/12065005/892f0726-afd8-11e5-8769-9656b318182f.PNG)

Screenshot of naginator cause when source is deleted.
![naginatorcausedeletedsource](https://cloud.githubusercontent.com/assets/1868117/12065006/8f1d13b2-afd8-11e5-9a80-bb8700c864a3.PNG)

ikedam commented:
Thanks for the update.

The project of the triggered build and that of the triggering build can always be assumed same.
Instead of storing `projectName` and `sourceBuildUrl`, 
* Add a member `project` to `NaginatorCause`. This should not be serialized and set to `transient`.
* Initialize `project` with `Cause#onAddedTo` and `Cause#onLoad`.

This allows links work even when the project is renamed.

litmanovich commented:
Thanks for the review. I changed NaginatorCause to have project as member and now it handles the case of renamed project.

ikedam commented on src/main/resources/com/chikli/hudson/plugin/naginator/NaginatorCause/description.jelly:
```diff
@@ -0,0 +1,19 @@
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+    <span>
+        <j:choose>
+            <j:when test="${app.getItemByFullName(it.projectName)!=null}">
```
You can simply use `it.project`.

ikedam commented on src/main/resources/com/chikli/hudson/plugin/naginator/NaginatorCause/description.properties:
```diff
@@ -0,0 +1,3 @@
+DescriptionWithLinkToSource=Started by Naginator after the failure of build <a class="model-link inside" href="{1}/{2}{3}">{0}</a>
```
Recommends trailing slashes for URLs of builds.

litmanovich commented:
Added new commit to use it.project and added trailing slash in href

ikedam commented:
Thanks for the update!
I tested this in my environment and saw this works almost perfect.

I found that this displays the text for deleted builds even for builds retried with an older version of naginator (it is caused for the change I requested, sorry).
I want it display `shortDescription` for those cases.

There are two ways to detect builds retried with older versions:
* `sourceBuildNumber` is less than 1.
* Change `sourceBuildNumber` from `int` to `Integer`. `null` is set for builds retried with older versions.

litmanovich commented:
I checked the build retry launched by old version of naginator - in my case the value of summary of removed build is `null`, so `shortDescription` shows: *Started by Naginator after build null failure*

I suggest in the case of either `it.project == null or it.sourceBuildNumber == null` show the following text: *Started by Naginator after the failure of previous build*. Do you think this will be Ok?

ikedam commented:
I suppose you tried naginator-plugin < 1.13.
`NaginatorCause#summary` was introduced naginator 1.13 and no care was taken for data created with older versions (I believe that was a bad manner).

Let's consider the compatibility only with naginator-plugin 1.13 - 1.16.1 for now.
We can expect `NaginatorCause#summary` is always set.

Adding a condition `it.summary == null` would be better if we want to take care also of naginator-plugin < 1.13.

litmanovich commented:
I added one more commit to check the value of `sourceBuildNumber`.

For builds launched by versions 1.13-1.16 it displays shortDescription

For builds launched by earlier versions of naginator it shows:
`Started by Naginator after the failure of previous build`

ikedam commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorCause.java:
```diff
@@ -9,15 +12,38 @@
+        return this.project.getUrl();
+    }
+
+    public int getSourceBuildNumber() { return this.sourceBuildNumber; }
```
This causes NPE for causes created with old versions, as this is compiled as `this.sourceBuildNumber.intValue()`.
Change the return type to `Integer`.

ikedam commented on src/main/resources/com/chikli/hudson/plugin/naginator/NaginatorCause/description.jelly:
```diff
@@ -0,0 +1,27 @@
+            </j:otherwise>
+        </j:choose>
+    </span>
+</j:jelly>
```
Please add a new line here.
I heard that text files should always end with new lines: http://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

litmanovich commented:
Sorry for not checking this in the first place. I now changed the return type of `getSourceBuildNumber` to `Integer` and added a newline at the end of jelly file.

ikedam commented:
Looks perfect!
Thanks a lot for the new feature.

ikedam commented:
I plan to make a new release after completing #26.
Please let me have some more time.

ikedam commented:
Sorry to have kept you wait for long time.
I released naginator-1.17 introducing this feature.
It will be available in the update center in a day.
